enhancements/0002-20221026-experimental-rest-transport-layer.md

```diff
@@ -0,0 +1,163 @@
+
+The motivation for this FED is, therefore, to make it easier for users to integrate, use, and deploy Flower in environments that have traditionally not been well supported.
+
+### Goals
+
+- *Increase infrastructure compatibility.* default configurations of common cloud infrastructure stacks.
+  - *Be compatible with caching infrastructure.*
+  - *Support CORS.*
+- *Increase client platform compatibility.* Flower clients should be able to run on platforms that are not well-supported by gRPC.
+- *Support proxy servers.* Enable client/server connections to go through proxy servers that may or may not alter the request, for example, to conceal the client's IP address from the server.
+- *Keep client identity optional.* Clients should be able to remain anonymous, client identity should be an optional component that users of the framework can opt for depending on the needs of their workload.
+- *Prevent clients from poisoning the model.*
+- *Limit server-side knowledge about individual clients.*
+
+### Non-Goals
+
+- *End-to-end latency.* The existing bidirectional stream between client and server offers stellar end-to-end latency (i.e., from scheduling a task for a client until that client starts executing the task locally). Most workloads do not require such low levels of latency.
+- *API stability.* The proposed REST API is not intended to be final.
+- *Client authentication.* There should be a forward-compatible path for adding client authentication, but client authentication does not need to be implemented in the first version.
+- *Compatibility with `start_server`.*
+- *Build a "RESTful" API.* The goal is to have a REST-inspired API, but not to build a fully idiomatic RESTful API.
+
+## Proposal
+
+Introduce a request/response REST transport layer. The REST transport layer would consist of a REST API server (integrated with the Flower server) and a REST client.
+
+## Design Details
+
+### Definition(s)
+
+Task: User-defined specification of what a client has to do, with the expectation that the client returns a result.
+
+### Protocol
+
+General idea:
+- Client always initiates conversation
+- Server responds in one of two ways:
+  - Reconnect at a later time.
+  - Execute a task and send the result.
+- Server has a (mild) form of control over the conversation by suggesting to the client when it should connect again. This can be used, among other things, to implement pace steering.
+
+Example:
+- Client connects to the server: "I'm available"
+- Server doesn't have any tasks: "reconnect in 5min"
+- Client sleeps for 5min, then connects again
+- Server say: here's a set of TASK-TOKEN-PAIRS, work on them and send me the RESULTS
+- Client receives the set of TASK-TOKEN-PAIRS, executes them locally (concurrently or sequentially, up to the client), and returns the RESULT (one after another or multiple results together, up to the client)
```

Comment:
I would make it optional for the client to execute the tasks. Maybe it's already like this, but from here it sounds like the server's response is an imperative and all assigned tasks have to be executed. 

The client might be unable to run a task because it doesn't have enough data for it or some other criteria doesn't check. But also it's a privacy advantage if there is some plausible deniability at the client side (i.e. it is not certain that a client will contribute to this round--straggler or not).

Oversampling might be enough to account for this potential "local rejection", or else the client could send a message/receipt to the server with the tasks it's going to run (in a more privacy preserving version we would add some noise to this _task execution promise_)

Reply:
Thanks @LorenzoMinto, good points, I updated the doc. Feel free to resolve this thread if you feel that the doc fully reflects our understanding.